Initial commit of minimal working functionality

fancyremarker opened the pull request:
A container launched from this built image will: 

* Run the Mirth Connect Administrator, listening for HTTP/HTTPS traffic on the default ports 80 and 443, respectively 
* Run a NGiNX server on port 3000 which proxies requests to Mirth Connect, forwaring to port 443 if it receives `X-Forwarded-Proto: https`

In this way, the image supports being run directly, or behind a sequence of reverse proxies (as is the case with Aptible Enclave Endpoints). Note that since Mirth Connect is listening internally on ports 80/443, the user- facing reverse proxy must also listen on ports 80/443, since Mirth Connect will insist on rewriting all URLs to include the ports it believes itself to be listening on. :sideeye:

By default, this image will use an in-container Derby database for storage of message data and configuration. It also supports using a PostgreSQL database, if a valid database URL is present in `$DATABASE_URL`.

cc: @UserNotFound @krallin 

There are a few more changes I'd like to make, for usability, but please let me know if you see anything totally off here.

krallin commented on templates/bin/mirthconnect-wrapper.sh:
```diff
@@ -0,0 +1,35 @@
+  fi
+
+  sed -i "s/^database =.*/database = postgres/" $PROPERTIES_FILE
+  sed -i "s/^database.url =.*/database.url = jdbc:postgresql:\/\/${host_and_port}\/${database}\?ssl=true\&sslfactory=org.postgresql.ssl.NonValidatingFactory/" $PROPERTIES_FILE
```
Note: you can use another separator (e.g. `:` instead of `/`)  to avoid needing to escape `/`

(not that you necessarily need it, but I thought I'd let you know)

We have some examples of this here: https://github.com/aptible/docker-postgresql/blob/8ce973fc2cde925a521a32d973ee11c425a96f54/bin/run-database.sh#L49-L54

fancyremarker replied:
Thanks! Done.

krallin commented on templates/etc/nginx/nginx.conf:
```diff
@@ -0,0 +1,69 @@
+  # Enable PFS for most browsers, and provide SSLv3 while mitigating POODLE
+  ssl_ciphers EECDH+ECDSA+AESGCM:EECDH+aRSA+AESGCM:EECDH+ECDSA+SHA384:EECDH+ECDSA+SHA256:EECDH+aRSA+SHA384:EECDH+aRSA+SHA256:EECDH+aRSA+RC4:EECDH:EDH+aRSA:RC4:!aNULL:!eNULL:!LOW:!3DES:!MD5:!EXP:!PSK:!SRP:!DSS:!CAMELLIA;
+  ssl_protocols SSLv3 TLSv1 TLSv1.1 TLSv1.2;
+  ssl_prefer_server_ciphers on;
```
We're not actually listening over SSL here - do we need these?

fancyremarker replied:
Nope! Artifact of borrowing this nginx.conf from elsewhere. Removed.

krallin commented on templates/bin/mirthconnect-wrapper.sh:
```diff
@@ -0,0 +1,35 @@
+
+# Launch Mirth Server
+echo "INFO: Launching Mirth Connect Administrator..."
+java -jar mirth-server-launcher.jar &
```
Note: Mirth appears to exit gracefully when signalled with `SIGTERM` (I tested this locally), but since it's not running as PID 1, it won't get the signal here. Do you think that's a problem?

fancyremarker replied:
Good call. Updated.

fancyremarker commented on LICENSE.md:
```diff
@@ -0,0 +1,22 @@
+Copyright (c) 2013 Aptible, Inc.
```
LOL, it was originally 2015 and I offset by 2 years in the wrong direction.

fancyremarker commented on Procfile:
```diff
@@ -0,0 +1 @@
+web: /usr/local/bin/mirthconnect-wrapper.sh
```
Good point. I don't think v1 support is too relevant/important in this case, so I can omit. I think that'd be the only reason to keep it around for a single-service app.

fancyremarker commented on Dockerfile:
```diff
@@ -0,0 +1,38 @@
+# See https://github.com/brandonstevens/mirth-connect-docker
+FROM java
```
👍 , done.

fancyremarker commented on Dockerfile:
```diff
@@ -0,0 +1,38 @@
+ENV MIRTHCONNECT_VERSION 3.5.0.8232.b2153
+
+# Install NGiNX
+RUN apt-get update && apt-get install nginx -y --no-install-recommends
```
Good call, done.

fancyremarker commented on Dockerfile:
```diff
@@ -0,0 +1,38 @@
+# Install Mirth Connect
+RUN \
+  cd /tmp && \
+  wget http://downloads.mirthcorp.com/connect/$MIRTHCONNECT_VERSION/mirthconnect-$MIRTHCONNECT_VERSION-unix.tar.gz && \
```
Updated to evaluate checksum, and also retrieve over HTTPS. Made an assumption about downloads.mirthcorp.com continuing to be hosted on S3, but I think that assumption is acceptable.

fancyremarker commented on Dockerfile:
```diff
@@ -0,0 +1,38 @@
+COPY templates/etc /etc
+COPY templates/bin /usr/local/bin
+
+CMD /usr/local/bin/mirthconnect-wrapper.sh
```
👍 , done.

fancyremarker commented:
Thanks again for the close review @krallin ! I've addressed all your points and have reworked mirthconnect-wrapper.sh a bit, in particular to make it as easy as possible for customers to set a non-default password *before* adding an Endpoint (see GIF). Also added clarifying steps to README.

![mirth-cli](https://user-images.githubusercontent.com/156030/27060250-83c9216e-4fa9-11e7-880f-f2c560c26dd8.gif)

krallin commented:
I very much approve of the new CLI mode! 👍

UserNotFound commented:
Works great for me in my locally!

fancyremarker commented:
Thanks, going to merge to be able to link to the repo instead of this PR. Still want to add built-in support for HTTP channels.

cronoh commented:
Worked great here, thanks!  I was also able to use the example code to get our HTTP channels working fine on Aptible as well. 👍

fancyremarker commented:
@cronoh : awesome, thanks for letting us know! We'll be adding some explicit code to help with configuring HTTP channels, and I'll ping you back when that's ready, in case it's helpful for you as a point of comparison.
